service/src/main/java/com/ge/predix/acs/attribute/readers/ExternalAttributeReader.java

```diff
@@ -92,20 +105,36 @@ private OAuth2RestTemplate getAdapterOauth2RestTemplate(final AttributeAdapterCo
             String adapterUrl = UriComponentsBuilder.fromUriString(adapterConnection.getAdapterEndpoint())
                     .queryParam(ID, identifier).toUriString();
 
+            AttributesResponse attributesResponse;
+
             try {
-                ResponseEntity<AttributesResponse> attributesResponse = adapter.getForEntity(adapterUrl,
-                        AttributesResponse.class);
-                attributes = attributesResponse.getBody().getAttributes();
+                attributesResponse = adapter.getForEntity(adapterUrl, AttributesResponse.class).getBody();
             } catch (final Throwable e) {
-                throw new AttributeRetrievalException(AttributeRetrievalException.getAdapterErrorMessage(adapterUrl),
-                        e);
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug(AttributeRetrievalException.getAdapterErrorMessage(adapterUrl), e);
+                }
+                attributesWrapper.setException(
+                        new AttributeRetrievalException(AttributeRetrievalException.getAdapterErrorMessage(adapterUrl),
+                                e));
+                return attributesWrapper;
+            }
+
+            if (attributesResponse.checkSizeLimit(maxNumberOfAttributes, maxSizeOfAttributesInBytes)
+                    == Integer.MAX_VALUE) {
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug(AttributeRetrievalException.getStorageErrorMessage(identifier));
+                }
+                attributesWrapper.setException(new AttributeRetrievalException(
+                        AttributeRetrievalException.getStorageErrorMessage(identifier)));
+                return attributesWrapper;
             }
+            attributesWrapper.setAttributes(attributesResponse.getAttributes());
```

Comment:
Combine this line with line 111 since that's the only place it seems like we need to set attributes:

`attributesWrapper.setAttributes(adapter.getForEntity(adapterUrl, AttributesResponse.class).getBody().getAttributes());`

If you do this, you don't need the `AttributesResponse` local variable.